From a81931603c25eee1ce4d178b58f5904d14e3a1d1 Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
Date: Thu, 16 Apr 2026 20:40:54 +0530
Subject: [PATCH] Revert "UPSTREAM:PROTOCOL: Reject duplicate per-link MACs across interfaces"

This reverts commit ee32af48a932ea0a442ef985d52b8b4df7ee40d1.

Reason for revert: prpl SLO STA failure

Patch-dependency:None

Change-Id: Ibfc0df60cf9e52efaac7e853f22b48723c5f5a52
Signed-off-by: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
---

diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index b29043f..d3c34ea 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -114,18 +114,6 @@ enum ieee80211_status_data {
 	IEEE80211_STATUS_SUBDATA_MASK	= 0x1ff0,
 };
 
-static inline int identical_mac_addr_allowed(int type1, int type2)
-{
-	return type1 == NL80211_IFTYPE_MONITOR ||
-	       type2 == NL80211_IFTYPE_MONITOR ||
-	       type1 == NL80211_IFTYPE_P2P_DEVICE ||
-	       type2 == NL80211_IFTYPE_P2P_DEVICE ||
-	       (type1 == NL80211_IFTYPE_AP && type2 == NL80211_IFTYPE_AP_VLAN) ||
-	       (type1 == NL80211_IFTYPE_AP_VLAN &&
-	       (type2 == NL80211_IFTYPE_AP ||
-	       type2 == NL80211_IFTYPE_AP_VLAN));
-}
-
 static inline bool
 ieee80211_sta_keep_active(struct sta_info *sta, u8 ac)
 {
diff --git a/net/mac80211/iface.c b/net/mac80211/iface.c
index c2f5fc5..644e6aa 100644
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -324,6 +324,18 @@ static int ieee80211_change_mac(struct net_device *dev, void *addr)
 	return _ieee80211_change_mac(sdata, addr);
 }
 
+static inline int identical_mac_addr_allowed(int type1, int type2)
+{
+	return type1 == NL80211_IFTYPE_MONITOR ||
+		type2 == NL80211_IFTYPE_MONITOR ||
+		type1 == NL80211_IFTYPE_P2P_DEVICE ||
+		type2 == NL80211_IFTYPE_P2P_DEVICE ||
+		(type1 == NL80211_IFTYPE_AP && type2 == NL80211_IFTYPE_AP_VLAN) ||
+		(type1 == NL80211_IFTYPE_AP_VLAN &&
+			(type2 == NL80211_IFTYPE_AP ||
+			 type2 == NL80211_IFTYPE_AP_VLAN));
+}
+
 static int ieee80211_check_concurrent_iface(struct ieee80211_sub_if_data *sdata,
 					    enum nl80211_iftype iftype)
 {
diff --git a/net/mac80211/link.c b/net/mac80211/link.c
index 64970e4..9e3abbc 100644
--- a/net/mac80211/link.c
+++ b/net/mac80211/link.c
@@ -256,62 +256,24 @@ static void ieee80211_free_links(struct ieee80211_sub_if_data *sdata,
 
 static int ieee80211_check_dup_link_addrs(struct ieee80211_sub_if_data *sdata)
 {
-	struct ieee80211_link_data *link, *other_link;
-	struct ieee80211_local *local = sdata->local;
-	struct ieee80211_sub_if_data *other_sdata;
-	const u8 *addr1, *addr2;
-	int link_id, o_link_id;
+	unsigned int i, j;
 
-	list_for_each_entry(other_sdata, &local->interfaces, list) {
-		bool is_same_sdata = (other_sdata == sdata);
+	for (i = 0; i < IEEE80211_MLD_MAX_NUM_LINKS; i++) {
+		struct ieee80211_link_data *link1;
 
-		if (!is_same_sdata && !ieee80211_sdata_running(other_sdata))
+		link1 = sdata_dereference(sdata->link[i], sdata);
+		if (!link1)
 			continue;
+		for (j = i + 1; j < IEEE80211_MLD_MAX_NUM_LINKS; j++) {
+			struct ieee80211_link_data *link2;
 
-		for (link_id = 0; link_id < IEEE80211_MLD_MAX_NUM_LINKS; link_id++) {
-			link = sdata_dereference(sdata->link[link_id], sdata);
-			if (!link)
+			link2 = sdata_dereference(sdata->link[j], sdata);
+			if (!link2)
 				continue;
 
-			addr1 = link->conf->addr;
-			/* skip unset address */
-			if (is_zero_ether_addr(addr1))
-				continue;
-
-			/* Compare link address against other_sdata's MLD address */
-			addr2 = other_sdata->vif.addr;
-			if (!is_same_sdata && !is_zero_ether_addr(addr2) &&
-			    ether_addr_equal(addr1, addr2) &&
-			    !identical_mac_addr_allowed(sdata->vif.type,
-							other_sdata->vif.type))
-				return -ENOTUNIQ;
-
-			for (o_link_id = 0;
-			     o_link_id < IEEE80211_MLD_MAX_NUM_LINKS;
-			     o_link_id++) {
-				/* Skip self and previously compared links
-				 * within same MLD
-				 */
-				if (is_same_sdata && o_link_id <= link_id)
-					continue;
-
-				other_link =
-					sdata_dereference(other_sdata->link[o_link_id],
-							  other_sdata);
-				if (!other_link)
-					continue;
-
-				addr2 = other_link->conf->addr;
-				/* skip unset address */
-				if (is_zero_ether_addr(addr2))
-					continue;
-
-				if (ether_addr_equal(addr1, addr2) &&
-				    (is_same_sdata ||
-				    !identical_mac_addr_allowed(sdata->vif.type,
-								other_sdata->vif.type)))
-					return -ENOTUNIQ;
-			}
+			if (ether_addr_equal(link1->conf->addr,
+					     link2->conf->addr))
+				return -EALREADY;
 		}
 	}
 